[quant][graphmode] FP16 quant support - Insert cast operators

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40710 [quant][graphmode] FP16 quant support - Operator Fusion
* **#40709 [quant][graphmode] FP16 quant support - Insert cast operators**
* #40708 [quant][graphmode] Add FP16 quant support - Insert Noop Observers

Summary:
Cast to kHalf and back to kFloat before the linear operator to mimic FP16 quant support

Test Plan:
python test/test_quantization.py test_convert_dynamic_fp16

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22335977](https://our.internmc.facebook.com/intern/diff/D22335977)

supriyar committed:
[quant][graphmode] FP16 quant support - Insert cast operators

Summary:
Cast to kHalf and back to kFloat before the linear operator to mimic FP16 quant support

Test Plan:
python test/test_quantization.py test_convert_dynamic_fp16

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graphmode] FP16 quant support - Insert cast operators"


Summary:
Cast to kHalf and back to kFloat before the linear operator to mimic FP16 quant support

Test Plan:
python test/test_quantization.py test_convert_dynamic_fp16

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization/helper.cpp:
```diff
@@ -578,11 +589,9 @@ bool is_module(
   const auto& match_vmap = match.values_map;
   Value* relu = match_vmap.at(vmap.at(vname));
   auto type = relu->type()->cast<ClassType>();
```
this can be removed?

jerryzh168 commented on torch/csrc/jit/passes/quantization/helper.h:
```diff
@@ -44,6 +44,9 @@ TORCH_API bool isScalar(Value* v);
 TORCH_API bool hitGraphInput(Value* value);
 
+// Return the module name that corresponds to the value.
+TORCH_API c10::optional<std::string> get_module_name(Value* value);
```
should we use the same naming convention like other functions? I just noticed this problem, probably the filter functions should follow the name convention as well. but we can do that later.

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -211,7 +248,10 @@ void insertQuantizationOps(
   Value* original_val = observer->input(1);
   Node *quant, *choose_qparams, *dequant;
-  if (quant_type == QuantType::DYNAMIC && !isWeight(module, observer_out)) {
+  if (quant_type == QuantType::DYNAMIC && isNoopObserver(observer->input(0))) {
+    dequant = insertFP16CastOps(g, observer_out);
+  } else if (
+      quant_type == QuantType::DYNAMIC && !isWeight(module, observer_out)) {
```
I feel it's more clear to merge the two branches into one? since the else branch is only for static I think.
if (quant_type == quanttype::dynamic) {
if (isNooposerver) {
} else if (!isWeight(..) {
}

} else {
...
}

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -211,7 +248,10 @@ void insertQuantizationOps(
+      quant_type == QuantType::DYNAMIC && !isWeight(module, observer_out)) {
     Value* dtype = g->insertGetAttr(self, qparam_names.back());
     std::tie(choose_qparams, quant, dequant) = insertChooseQParamQuantDequant(
         g, observer_out, dtype, at::Symbol::aten(quantize_func));
```
probably better to add a comment for else branch, it's accepting both dynamic + activation and static quant

jerryzh168 commented on test/quantization/test_quantize_jit.py:
```diff
@@ -2710,6 +2710,27 @@ def forward(self, x):
+                return self.fc(x)
+
+        m = torch.jit.script(M())
+        m = prepare_dynamic_jit(m, {'': float16_dynamic_qconfig})
+        m = convert_dynamic_jit(m, debug=True)
```
this can be done in one API?

supriyar committed:
Update on "[quant][graphmode] FP16 quant support - Insert cast operators"


Summary:
Cast to kHalf and back to kFloat before the linear operator to mimic FP16 quant support

Test Plan:
python test/test_quantization.py test_convert_dynamic_fp16

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graphmode] FP16 quant support - Insert cast operators"


Summary:
Cast to kHalf and back to kFloat before the linear operator to mimic FP16 quant support

Test Plan:
python test/test_quantization.py test_convert_dynamic_fp16

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
